### DISCUSSION TITLE
MultiApp Transfer to nearest node only

### DISCUSSION THREAD

**Discussion by WilkAndy** [idaholab/moose#17806] [Q&A General] [answered]
Hi everyone,

@lynnmunday , @cbolisetti , @rpodgorney , KT and I are working on simulating flow through fractured porous media using a MultiApp approach.  In this work, the "fracture" is a 2D mesh living within the "matrix", which is a 3D mesh.  The fracture nodes/elements don't usually conform with the matrix nodes/elements.

I can pass information from the matrix to the fracture using, eg, MultiAppNearestNodeTransfer or MultiAppInterpolationTransfer.  These Transfers put a value at each node of the fracture mesh from the nearest node of the matrix mesh (or interpolate a few matrix values in the case of InterpolationTransfer).

I need to pass information from the fracture to the matrix.  Physically, heat is flowing from the fracture to the matrix.  Each node of the fracture should pass information (heat) to **one node of the matrix only** : the matrix node nearest to the fracture node.  Using a MultiAppNearestNodeTransfer does not achieve this, for **all** nodes of the matrix get a value (heat) from their nearest fracture node.  There are other possibilities that would also work, for instance, passing information from each of the fracture nodes to nodes of the matrix element that contains the fracture node.  I also have to be careful that many fracture nodes may be close to a single matrix node (in which case the transferred thing - heat energy - would have to be summed, just like putting lots of DiracKernels inside a single element).  But before considering these subtleties, i wanted to ask:

Does MOOSE or one of its Apps have anything like this, or can you see another approach that would work ?

a

**Comment by vincentlaboure** [comment]
Would `MultiApps/*/bounding_box_padding` help with your need?

**Reply by WilkAndy** [reply]
Thanks @vincentlaboure - this might be the kind of thing i'm looking for.  I couldn't get it to work as desired in simple cases, but, regardless, in complicated-mesh cases, i suspected the "padding" and "inflation" would have to be dependent on the matrix mesh.  Eg, some of the fracture nodes might be very close to a matrix node, meaning the padding&inflation needs to be small for them, while other fracture nodes would need more padding&inflation because they're a long distance from any matrix node

**Reply by mcacace** [reply]
@WilkAndy I had a similar (though different in a way) problem, need to compute and pass back and forth traction stresses from matrix elements to frac elements and the other way around (slip modelling). What I did was to compute (via an UO, but can be done in a Multiapp as well) a map among those and use that for passing the information.  In my case all is comforming, that is there are max 2 3D elements per each 2D elements, and the geometry does not change, but the map could be extended to more complicated cases (I fill in vectors of ids) and for evolving geometries (by running the map during run time and not on initialization as I am doing). The only caveat here is that my mapping is not 100% thread safe and so far works only on serial meshes (did not have enough time/man power to generalize it). In case I could share my piece of code (would be lovely to have it smarten). Hope this helps, mauro

**Reply by WilkAndy** [reply]
Thank @mcacace , that could be good.  The team is meeting in about 12 hours and we'll discuss this in detail - i'll get back to you if we end up needing this feature.

**Comment by hugary1995** [comment, marked as answer]
Does DiracKernel work? It seems you have considered it but I am curious why it doesn't satisfy your needs.

**Reply by lynnmunday** [reply]
We are using the porousFlowHeatMassTransfer kernel where we apply the temperature difference between the fracture and matrix times a heat transfer coefficient.  It is easy to pass the temperature back and forth between the apps but if we use the diracKernel, we would need to compute a heat flux and pass that.  Maybe we could compute that from the temperature difference in the same way. Would you suggest making a diracKernel for each node on the subapp and passing that point along with a flux to the main app using a reporter?

**Reply by hugary1995** [reply]
Lynn, it doesn't have to be all that complicated. I took a look at `PorousFlowHeatMassTransfer`, basically all you want is a coupled variable `v` which is temperature or whatever information from the lower dimensional domain, right? There is a very simple trick to do that. It is very similar to an L2-projection, but you replace the source term with discrete approximations using DiracKernel.

I quickly put together a test in moose, and it is performing well. I first manufacture a field (suppose that is the information from the lower dimensional domain), then I sample the field value at N points and reconstruct the field onto the matrix mesh using this trick. Some results are here:

![comparison](https://user-images.githubusercontent.com/31555204/118200065-569dc500-b422-11eb-8cc5-152c5b79f2ab.png)

I am only using 20 elements in the matrix mesh, and with 64 sampling points from the lower dimensional domain, the error is already negligible. That corresponds to 2~3 points per element, which is pretty cheap. If you think this could work, we can discuss offline.

**Reply by hugary1995** [reply]
Per our discussion, this is input file I used:

```
[Mesh]
  [gmg]
    type = GeneratedMeshGenerator
    dim = 2
    xmax = 20
    ymax = 1
    nx = 20
    ny = 1
  []
[]
[Variables]
  [T]
  []
[]
[Kernels]
  [react]
    type = Reaction
    variable = T
  []
[]
[DiracKernels]
  [projection]
    type = VectorPostprocessorPointSource
    variable = T
    vector_postprocessor = data
  []
[]
[VectorPostprocessors]
  [data]
    type = CSVReader
    csv_file = data_${N}.csv
    outputs = none
  []
  [T]
    type = LineValueSampler
    variable = T
    start_point = '0 0.5 0'
    end_point = '20 0.5 0'
    num_points = 200
    sort_by = x
  []
[]
[Executioner]
  type = Steady
[]
[Outputs]
  file_base = T_${N}
  [csv]
    type = CSV
    execute_vector_postprocessors_on = FINAL
  []
[]
```

**Reply by lynnmunday** [reply]
@WilkAndy I think @hugary1995 method will be really easy to try out and I think its probably going to work better than a nodal transfer

**Reply by WilkAndy** [reply]
How is this going to work for complicated fractures?  As i understand it, each fracture node becomes a DiracKernel in the matrix.  So it looks like you have to pre-define the position of each fracture (or fracture node), which will be difficult and annoying for users.  Also difficult when mesh adaptivity is used?

**Reply by hugary1995** [reply]
No, you only need to sample temperature at each quadrature points in the fracture mesh, then transfer the VPP to the main app. In main app, you only need ONE DiracKernel.

**Reply by WilkAndy** [reply]
How do you sample the temperature at each qp of the fracture?  I'm thinking of a fracture with crazy geometry that can't be written as a simple function, so the qps will be thousands/millions of "random" points in space.  I think something is missing from my understanding - perhaps i don't understand how the matrix app knows where those points qps are.

**Reply by hugary1995** [reply]
Yeah we should back off a little bit... 

In your original post, you were mentioning a 2D mesh representing "fracture". I was assuming there is a temperature field on your fracture mesh, and you are using the difference between the temperature on the fracture and the temperature on the matrix mesh to drive some transfer (which happens in your main app, i.e. on the matrix mesh).

Did I understand something wrong?

**Reply by WilkAndy** [reply]
This is correct.  The difficulty we're struggling with is that the 2D fracture physically transfers heat energy only to the 3D matrix elements that contain the 2D fracture nodes.  We can do that in two ways

- `Transfer` heat energy from the 2D fracture to the appropriate 3D matrix elements
- `Transfer` temperature information from the 2D fracture to the appropriate 3D matrix elements, and (this is the easy bit) let the matrix app figure out how much heat-energy that corresponds to.

I believe your approach would `Transfer` the information to an appropriate `DiracKernel` in the matrix app.  But i don't know how this `DiracKernel` or `VectorPostprocessor` knows the positions of the fracture nodes or quadpoints.  What am I missing?

**Reply by hugary1995** [reply]
Okay we are close. In the subapp (fracture mesh), you need to have a VPP or reporter that loop over all elements, sample the field (be it temperature or heat energy, or whatever) at each quadrature point (scaled appropriately with element volume), and store the location of the qp together with the sampled value in the VPP with the structure like:
```
x,      y,       z,      value
2.85,   0.52,    0.02,   0.826
2.11,   3.13,    9.72,   11.96
......
```
Then you can use `MultiAppVectorPostprocessorTransfer` to transfer the data of the VPP from the subapp to the main app.
In the main app, you will use a structure similar to the input file I pasted above to reconstruct the sampled field from the subapp.

In the main app, you always only need one `VectorPostprocessorPointSource` DiracKernel. This DiracKernel will read information from the VPP, apply point sources at all points that you provided.

I think it'll actually be easier to just sample and transfer the heat energy. Because, if you were to transfer temperature from fracture mesh to matrix mesh, you will end up with mostly zero temperature in the elements that do not intersect with the fracture mesh. You can probably do some smart things about that though, like only evaluating your `PorousFlowHeatMassTransfer` kernel on elements that intersect with the fracture mesh? But anyways, choosing what to transfer from subapp to main app is up to you, what I am describing is a general way of transferring information.

**Reply by WilkAndy** [reply]
Yes, this will work, thank you @hugary1995 .  I agree about transferring heat energy - this is what we really want to do.

@hugary1995 - do you have any comments about performance?  Eg, if the fracture mesh doesn't change then obviously the `DiracKernel` can cache positions of the points, which could improve performance significantly.  Will the `DiracKernel` do this automatically?

@lynnmunday - i'll leave it to you to figure out whether this is easier than  #17835.   The VPP or Reporter actually needs to store information about the nodal positions, not the quadpoint positions.  The "information" will come from a `save_in` Variable of the `PorousFlowHeatMassTransfer` Kernel, so will be measured in Joules/second (or kg/s for mass flow), hence, there is no need to scale by any element volumes.  That is you've got a nodal AuxVariable, and you want to store (x, y, z, value) for all the nodes of the fracture mesh (where "value" is the value of the AuxVariable at the node).

@cbolisetti - now we've got two potential methods to solve our problem.  I just want to mention that both will be subject to finite-element discretisation errors, eg, if your matrix has a coarse mesh then the errors will be larger than a fine mesh.

**Reply by hugary1995** [reply]
Yes DiracKernel does caching by default. But I don't know the underlying logic for caching, so I can't say much about performance in your case. I would be surprised if it turns out to be slow. Maybe framework developers can comment here.

Even if it doesn't do caching at all, I believe you can speed things up by telling the DiracKernel which element a point belongs to. Doing a geometric search in the fracture mesh while writing data to the VPP will be much much faster than a geometric search in the matrix mesh.

**Reply by WilkAndy** [reply]
Hang on, maybe i don't understand this properly yet - how does the fracture app know anything about the matrix mesh (to do the geometric search)?

**Reply by hugary1995** [reply]
> Even if it doesn't do caching at all, I believe you can speed things up by telling the DiracKernel which element a point belongs to. Doing a geometric search in the fracture mesh while writing data to the VPP will be much much faster than a geometric search in the matrix mesh.

This was pure guessing. If you don't know the matrix mesh in the subapp then nevermind me.

**Reply by WilkAndy** [reply]
Aha, good, thanks @hugary1995

**Reply by WilkAndy** [reply]
Hey @hugary1995 - i just wanted to say thanks for all your input here.  I love being part of such a friendly and helpful team!

**Reply by cbolisetti** [reply]
Hi @WilkAndy, @hugary1995, and @lynnmunday the DiracKernel option sounds good too and might involve less code development than #17835. I guess the question is which method scales well since we might have a number of fracture nodes that are in the 1e5 - 1e6 scale. Do you know if `VectorPostProcessorPointSource` is parallelized? I couldn't see anything in the code, so I am not sure if something is being done in the base class.

**Reply by cbolisetti** [reply]
And yes, @hugary1995, thanks so much for the input!

**Reply by WilkAndy** [reply]
I agree about parallelism, @cbolisetti - we (@lynnmunday ) will have to be careful with parallelism - it is very likely that most of the fracture nodes will be on different processors than their containing matrix elements.  But @17835 also has nontrivial parallelism, but there @lynnmunday can probably copy bits of code from other transfers.

**Comment by WilkAndy** [comment]
Let's take further discussion to issues #17854 or #17835, whichever is appropriate
